Step3 로또(자동) PR

develo-pyo opened the pull request:
안녕하세요. 강태화님.
Step3 로또(자동) PR 드립니다.

기능단위 커밋과 TDD 개발이 아직 어렵네요..
처음에 TDD로 구현을 하다가도 정신차리고 보면 Test코드 없이 프로덕션코드를 우선 개발하고, 뒤늦게 Test 코드를 작성하고 있네요.
도중에 커밋도 치지 않은채 개발을 쭉 하고 있다가 뒤늦게 정신차리면 거의 개발이 다 되어있고...
더 많은 연습이 필요할 것 같습니다.

사용자 입력값에 대한 유효성검사는 요구사항에 없었는데 불필요한 개발을 한건 아닌가 모르겠네요.

리뷰 부탁드립니다.
감사합니다.

develo-pyo committed:
docs: 로또 요구사항 정리

develo-pyo committed:
docs: step3 로또(자동) 요구사항 정리

develo-pyo committed:
feat: 로또 도메인 구현

중복되지않는 1~45 숫자 중 6개를 갖는 로또 도메인
테스트 코드 구현
1) 중복되지 않은 로또 번호 6개 생성 테스트
2) 적중한 로또번호 갯수 리턴 테스트

develo-pyo committed:
feat: 로또결과 도메인 구현

1. 로또결과(각각의 등수) 메모리에 저장할 수 있는 일급객체
2. 수익률 계산 메소드
3. 테스트 구현
  1) 로또 N개의 각 등수가 총 몇개인지 리턴 테스트
  2) 수익률 계산 테스트

develo-pyo committed:
feat: 로또게임 도메인 구현

1. 구매한 금액으로 구매가능한 로또 갯수만큼 로또 생성하는 로또게임 도메인 구현
2. 로또 당첨번호를 입력받아 로또 N개의 결과 생성 메소드 구현
3. 테스트 구현
  1) 구매 가능 로또 갯수 테스트

develo-pyo committed:
feat: view 구현

1. 입력값 처리 view 구현
2. 출력값 처리 view 구현

develo-pyo committed:
feat: 로또메인 구현

develo-pyo committed:
refactor: 도메인 패키지 경로 분리

도메인 클래스 도메인 패키지 경로로 이동

develo-pyo committed:
refactor: 상수값 분리

상수값 분리

develo-pyo committed:
refactor: 상수값 분리

상수값 분리

develo-pyo committed:
refactor: 상수값 분리

상수값 분리

develo-pyo committed:
feat: 사용자 입력값 유효성검사 구현

1.사용자 입력값 유효성 검사 구현
2.테스트 구현

develo-pyo committed:
refactor: lottos 일급콜렉션 사용

lotto 리스트를 인스턴스로 갖는 lottos 일급콜렉션 사용하여 리펙토링

develo-pyo committed:
refactor: 불필요 import 제거, 변수명 변경

1.불필요 import 제거
2.*winner*->*winning* 으로 변수명 변경

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+    private static final int SUBLIST_START_IDX = 0;
+    private static final int SUBLIST_END_IDX = 6;
+
+    public Lotto(){
```
```suggestion
    public Lotto() {
```

자동 포멧팅세팅 해주세요 :pray:

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+    public int countCollectNumber(Lotto winningNumbers) {
+        int result = 0;
+        for(int winningNumber : winningNumbers.lottoNumbers){
+            if(lottoNumbers.contains(winningNumber)){
```
## 프로그래밍 요구사항
- indent(들여쓰기) depth를 2단계에서 1단계로 줄여라.
  - depth의 경우 if문을 사용하는 경우 1단계의 depth가 증가한다. if문 안에 while문을 사용한다면 depth가 2단계가 된다.
----

`for`안에 `if`도 2단계 아닐까 생각이되네요 :thinking: 

`depth`를 줄일수 있을지 고민해보는것도 좋을것 같아요!

wooobo commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+public class Lottos {
+
+    private final List<Lotto> lottos;
+    private final LottoResult lottoResult;
```
```
- 규칙 8: 일급 콜렉션 사용 
  - 콜렉션을 포함한 클래스는 반드시 다른 멤버 변수가 없어야 한다. 
```
일급콜렉션에 대해 좀 더 고민해보시면 좋을것같아요~

wooobo commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public void printLottoResult(){
+        OutputView.printLottoResult(lottoResult);
```
`domain`, `view` 레이어를 분리해주셨네요 :+1: 

하지만, 중요한 `domain`안에 `view`에 대한 의존성이 있는것이 맞을지 의문이에요 :thinking: 
좀 더 책임과 역할을 분리해보면 어떨까요?

wooobo commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,33 @@
+        for(int i=MIN_WINNING_NUM; i< PRIZE_MONEY.length; i++){
+            System.out.printf(OUT_MESSAGE_LOTTO_RESULT, i, PRIZE_MONEY[i], lottoResult.getLottoResult(i));
+        }
+        System.out.printf(OUT_MESSAGE_PROFIT_RATIO, lottoResult.calculateProfitRatio());
```
```
지난 주 당첨 번호를 입력해 주세요.
1,2,3,4,5,6
당첨 통계
----------
3개 일치(5000)- 1개
4개 일치(50000)- 0개
5개 일치(1500000)- 0개
6개 일치(2000000000)- 0개
총 수익률은 0.35 입니다.
```

`총 수익률은 0.35입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)` 
`수익률 1`미만에 대한 출력 메시지가 누락된것같아요!

wooobo replied:
```
지난 주 당첨 번호를 입력해 주세요.
1,2,6,4,8,4
```
당첨번호에 대한 중복검증이 필요할것 같아요!

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+        return result;
+    }
+
+    public List<Integer> generateUniqueLottoNumbers(){
```
## 프로그래밍 요구사항

`method가 한 가지 일만 하도록 최대한 작게 만들어라.`

----

하나의 메서드에 여러 기능이 구현되어 있는것 같아요!
메서드에서 한가지 일만하도록 개선해보면 어떨까요 :pray:

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+public class Lotto {
+
+    private final List<Integer> lottoNumbers;
+    public static final int MIN_LOTTO_NUMBER = 1;
```
아래 순서가 일반적으로 사용하는 순서라고 알고있어요!
참고해주세요

```
class A {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

wooobo commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public int calculateLottoAmount(){
+        return payAmount/LOTTO_PRICE;
```
`private final int payAmount;`
기본 타입이 아니라  `PayAmount` 라는 객체가 있다면 어떨까요?

그렇다면, `PayAmount` 에게 해당 로직을 계산하는 책임을 부여할수 있을것 같아요 :thinking:

wooobo commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,36 @@
+        return lottoResult.getOrDefault(collectNumberCnt, 0);
+    }
+
+    public double calculateProfitRatio(){
```
수익률을 계산하는 책임을 가지는 객체로 분리시켜보아도 좋을것 같네요 :)

wooobo reviewed (changes requested):
3단계 잘 구현해주셨습니다! :)


코멘트 남겨드렸는데 확인 후 
재요청 부탁드릴께요! :pray:

wooobo commented:
> 사용자 입력값에 대한 유효성검사는 요구사항에 없었는데 불필요한 개발을 한건 아닌가 모르겠네요.

유효성검사는 `view` or `domain` 이든 어딘가에는 있어야하지 않을까 생각되네요!
다만, 유효성 검사를 어떤 레이어에 부여될지가 중요할것 같아요! `(책임과 역할)`


> 기능단위 커밋과 TDD 개발이 아직 어렵네요..

TDD가 익숙하지 않다면, 테스트 코드부터 하나하나 만들어가는게 쉽지 않다고 생각해요!
이번 우테크 pro 과정의 미션을 하나하나 진행하시다 보면 많은 도움이 되실꺼에요!

develo-pyo committed:
feat: 중복값 검사 메소드 구현

1.중복값 검사 메소드 구현
2.중복값 검사 메소드 테스트 구현

develo-pyo committed:
fix: 유저입력 뷰 내에서 중복값 검사 로직 추가

develo-pyo committed:
refactor: 로또 번호 생성 메소드 별도 클래스 분리

1. 로또번호 생성 메소드 별도 클래스 분리
2. 로또번호 일치 갯수 판별 메소드 리팩토링(indent 2->1)
2. 로또번호 생성 테스트

develo-pyo committed:
refactor: 수익률 계산 메소드 별도 클래스로 분리

1. 수익률 계산 메소드 별도 클래스로 분리
2. 수익률 계산 메소드 테스트 분리

develo-pyo committed:
refactor: 구입금액 별도 클래스로 분리

1. 구입금액 별도 클래스로 분리
2. 구입금액 별 구매가능로또 갯수 테스트 분리

develo-pyo committed:
refactor: 컨트롤러 기능별로 메소드 분리

1. 컨트롤러 기능별로 작은 단위로 메소드 분리
2. 컨트롤러 패키지 경로로 이동
3. 경로변경에 따른 import 변경
4. 불필요 테스트 삭제

wooobo commented on src/main/java/lotto/util/ProfitCalculator.java:
```diff
@@ -0,0 +1,20 @@
+    public static final int[] PRIZE_MONEY = {0, 0, 0, 5000, 50000, 1500000, 2000000000};
+    public static final int MIN_WINNING_NUM = 3;
+
+    public static double calculateProfit(LottoResult lottoResult, int lottoAmount) {
```
## 프로그래밍 규칙
`규칙 3: 모든 원시값과 문자열을 포장한다.`

--- 
[[Java] 원시값 포장이란?](https://livenow14.tistory.com/37)
참고해주세요~

wooobo commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -14,29 +14,27 @@ public class InputView {
 
-    public static List<Integer> inputLottoWinningNumbers(){
-        Scanner sc = new Scanner(System.in);
+    public static List<Integer> inputLottoWinningNumbers() {
```
프로그래밍 요구사항에  `메소드의 크기가 최대 10라인을 넘지 않도록 구현한다.` 
해당 부분에 대해서 고민해주시면 좋을것 같아요!

wooobo commented on src/test/java/lotto/util/InputValidatorTest.java:
```diff
@@ -3,33 +3,45 @@
     @Test
-    @DisplayName("6개의 로또 숫자가 아닌 경우 검사")
-    public void validateLottoNumberCountTest(){
+    @DisplayName("당첨번호 중복 테스트")
```
테스트를 좀 더 강화해보셨네요 👍

wooobo reviewed (approved):
3단계 진행하느라 수고 많으셨습니다!

추가적인 코멘트 소소하게 달아두었습니다 확인부탁드립니다~

이만 `merge` 하겠습니다!

4단계도 잘부탁드립니다~ 👍
